digital_printing/digital_printing/doctype/print_order/print_order.js

```diff
@@ -0,0 +1,221 @@
+			}
+		});
+	}
+
+	on_upload_complete() {
+		return this.get_items_from_attachments();
+	}
+
+	default_gap() {
+		this.override_default_value_in_items('design_gap');
+	}
+
+	default_qty() {
+		this.override_default_value_in_items('qty');
+	}
+
+	default_uom() {
+		this.override_default_value_in_items('uom');
+		if (this.frm.doc.default_uom == "Panel") {
+			this.frm.set_value("default_qty_type", "Print Qty");
+		} else {
+			this.frm.set_value("default_length_uom", this.frm.doc.default_uom);
+		}
+	}
+
+	default_qty_type() {
+		this.override_default_value_in_items('qty_type');
+	}
+
+	default_wastage() {
+		this.override_default_value_in_items('per_wastage');
+	}
+
+	default_length_uom() {
+		this.override_default_value_in_items('length_uom');
+	}
+
+	items_add(doc, cdt, cdn) {
+		this.set_default_values_in_item(cdt, cdn);
+	}
+
+	before_items_remove(doc, cdt, cdn) {
+		let row = frappe.get_doc(cdt, cdn);
+		let file_name = this.frm.attachments.get_file_id_from_file_url(row.design_image);
+		this.frm.attachments.remove_attachment(file_name);
+		this.calculate_totals();
+	}
+
+	design_image(doc, cdt, cdn) {
+		var me = this;
+		let row = frappe.get_doc(cdt, cdn);
+
+		return frappe.call({
+			method: "get_image_details",
+			args: {
+				image_url: row.design_image
+			},
+			doc: me.frm.doc,
+			callback: function(r) {
+				if (!r.exc && r.message) {
+					for (const [key, value] of Object.entries(r.message)) {
+						frappe.model.set_value(cdt, cdn, key, value);
+					}
+				}
+			}
+		});
+	}
+
+	design_gap() {
+		this.calculate_totals();
+	}
+
+	qty() {
+		this.calculate_totals();
+	}
+
+	uom(doc, cdt, cdn) {
+		let row = frappe.get_doc(cdt, cdn);
+
+		if (row.uom == 'Panel') {
+			frappe.model.set_value(cdt, cdn, "qty_type", "Print Qty");
+		} else {
+			frappe.model.set_value(cdt, cdn, "length_uom", row.uom);
+		}
+		this.calculate_totals();
+	}
+
+	qty_type() {
+		this.calculate_totals();
+	}
+
+	per_wastage() {
+		this.calculate_totals();
+	}
+
+	length_uom() {
+		this.calculate_totals();
+	}
+
+	override_default_value_in_items(cdf) {
+		(this.frm.doc.items || []).forEach(d => {
+			this.set_default_values_in_item(d.doctype, d.name, cdf);
+		});
+	}
+
+	set_default_values_in_item(cdt, cdn, cdf=null) {
+		let defaults = {
+			'design_gap': this.frm.doc.default_gap,
+			'qty': this.frm.doc.default_qty,
+			'uom': this.frm.doc.default_uom,	
+			'qty_type': this.frm.doc.default_qty_type,
+			'per_wastage': this.frm.doc.default_wastage,
+			'length_uom': this.frm.doc.default_length_uom,
+		}
+
+		if (cdf) {
+			if (defaults[cdf]) {
+				frappe.model.set_value(cdt, cdn, cdf, defaults[cdf]);
+			}
+		} else {
+			for (const [key, value] of Object.entries(defaults)) {
+				if (value) {
+					frappe.model.set_value(cdt, cdn, key, value);
+				}
+			}
+		}
+	}
+
+	calculate_totals = () => {
+		this.frm.doc.total_print_length = 0;
+		this.frm.doc.total_fabric_length = 0;
+		this.frm.doc.total_panel_qty = 0;
+
+		let conversion_factors = {
```

Comment:
Maybe you can move it outside the function .. make it a constant attribute of the controller?